backupccl: set DescIDGenerator directly in full-cluster restore

pbardea opened the pull request:
The DescIDGenerator can be set directly rather than incremented many
times by using the underlying increment function.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45866)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -194,19 +194,14 @@ func allocateTableRewrites(
-			if err != nil {
-				return nil, err
-			}
+		numberOfIncrements := int64(maxDescIDInBackup - uint32(sql.MaxDefaultDescriptorID))
```
If we need to lock out full-cluster restore when there are existing descriptors GC'ing or something, we should do that elsewhere, but this code here is responsible for ensuring `keys.DescIDGenerator` is exactly `maxDescIDInBackup` -- i'd feel a little more confident that it was doing that if we just set it directly to that value, of, if we want to inc by a delta, read the existing value instead of _assuming_ it is `sql.MaxDefaultDescriptorID`

dt commented:
we just decided we can put this key to the value we want, right?

pbardea commented:
Yep, I was updating it to just do a set, but the tests were failing for some reason. Today I found that it was likely due to #49124, which doesn't clean up schema entries properly. So when we try and re-use database IDs that were previously dropped in the restoring cluster we get errors. Ideally the fix for that can be merged soon so this can quickly go on top of that. Otherwise we could just keep the DescIDGenerator to be max(maxDescIDInBackup, currentDescIDGenerator). This would mean that after the restore you may skip a bunch of IDs, but I think that generally should be OK.

pbardea committed:
backupccl: set DescIDGenerator directly in full-cluster restore

The DescIDGenerator can be set directly rather than incremented many
times by using the underlying increment function.

Release note: None

pbardea commented:
This is RFAL, the changes to `pkg/sql/drop_database.go` is a temporary bandaid until #49124 finishes merging.

pbardea commented:
TFTR!
bors r=dt

pbardea commented:
bors r=dt
